docs/codeql/codeql-language-guides/using-api-graphs-in-ruby.rst

```diff
@@ -0,0 +1,185 @@
+to methods on those instances. The following snippet finds instantiations of the ``Regexp`` class:
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("Regexp").getInstance()
+
+Note that the ``getInstance`` method also includes subclasses. For example if there is a 
+``class SpecialRegexp < Regexp`` then ``getInstance`` also finds ``SpecialRegexp.new``.
+
+The following snippet builds on the above to find calls of the ``Regexp#match?`` instance method:
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("Regexp").getInstance().getMethod("match?")
+
+Subclasses
+----------
+
+Many libraries are used by extending one or more library classes. To track this
+in the API graph, you can use the ``getASubclass`` method to get the API graph node corresponding to
+the immediate subclasses of a node. To find *all* subclasses, use ``*`` or ``+`` to apply the
+method repeatedly. You can see an example where all subclasses are identified using ``getASubclass*`` below.
+
+Note that ``getASubclass`` can only return subclasses that are extracted as part of the CodeQL database 
+that you are analyzing. When libraries have predefined subclasses, you will need to explicitly include them in your model.
+For example, the ``ActionController::Base`` class has a predefined subclass ``Rails::ApplicationController``. To find
+all subclasses of ``ActionController::Base``, you must explicitly include the subclasses of ``Rails::ApplicationController`` as well.
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+
+    API::Node actionController() {
+      result =
+        [
+          API::getTopLevelMember("ActionController").getMember("Base"),
+          API::getTopLevelMember("Rails").getMember("ApplicationController")
+        ].getASubclass*()
+    }
+
+    select actionController()
+
+
+Using the API graph in dataflow queries
+---------------------------------------
+
+Dataflow queries often search for points where data from external sources enters the code base
+as well as places where data leaves the code base. API graphs provide a convenient way to refer 
+to external API components such as library functions and their inputs and outputs. 
+However, you do not use API graph nodes directly in dataflow queries. 
+
+- API graph nodes model entities that are defined outside your code base.
+- Dataflow nodes model entities defined within the current code base. 
+
+You bridge the gap between the entities outside and inside your code base using 
+the API node class methods: ``asSource()`` and ``asSink()``.
+
+The ``asSource()`` method is used to select dataflow nodes where a value from an external source
+enters the current code base. A typical example is the return value of a library function such as
+``File.read(path)``:
+
+.. code-block:: ql
+
+    import codeql.ruby.ApiGraphs
+
+    select API::getTopLevelMember("File").getMethod("read").getParameter(1).asSource()
```

Comment:
```suggestion
    select API::getTopLevelMember("File").getMethod("read").getReturn().asSource()
```
Looks like you copy-pasted a bit much from the below example.

Reply:
Thank you, well spotted!